### PR TITLE
feat: introduce group manager

### PR DESCRIPTION
**Pull request by kian99**
## Description

This PR follows on from https://github.com/canonical/jimm/pull/1455 and create a group manager.

Moved group methods to group package.
Moved createTestControllerEnvironment function to jimmtest package.

Fixes [JUJU-7245](https://warthogs.atlassian.net/browse/JUJU-7245)

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [x] Covered by unit tests
- [x] Covered by integration tests

[JUJU-7245]: https://warthogs.atlassian.net/browse/JUJU-7245?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/jimm/group/group.go
Comment (alesstimec) [inline internal/jimm/group/group.go:4]: new line (+1 for my comment count \o/)

Comment (alesstimec) [reply internal/jimm/group/group.go:4]: also, as the package needs a godoc (other than copyright)

File: internal/jimm/group/group.go
Comment (alesstimec) [inline internal/jimm/group/group.go]: hmm.. slightly worried that this is not done in a transaction. while i don't expect us to have much load on the group manger for this to become a problem i still think it should be done in a transaction.

File: internal/jimm/group/group.go
Comment (alesstimec) [inline internal/jimm/group/group.go]: same here.. transaction, please.

File: internal/jimm/jimm.go
Comment (alesstimec) [inline internal/jimm/jimm.go:95]: does the `GroupManager` need to be exported, or could we use the `GroupManager()` getter method instead?

File: internal/testutils/jimmtest/fixture.go
Comment (alesstimec) [inline internal/testutils/jimmtest/fixture.go:3]: you know..

File: internal/testutils/jimmtest/fixture.go
Comment (alesstimec) [inline internal/testutils/jimmtest/fixture.go:35]: this needs to return a struct type. there are way to many returns

**Review by alesstimec — CHANGES_REQUESTED** [review]

File: internal/jimm/group/group.go
Comment (ale8k) [inline internal/jimm/group/group.go]: s/permission/group

File: internal/jimm/jimm.go
Comment (kian99) [reply internal/jimm/jimm.go:95]: It doesn't need to be exported right now, but the plan is to remove the GroupManager() getter as soon as we can. Up to you whether you want me to make it exported or not right now.

File: internal/testutils/jimmtest/fixture.go
Comment (kian99) [reply internal/testutils/jimmtest/fixture.go:3]: I wonder if this isn't the linter always adding this copyright header like this.

File: internal/testutils/jimmtest/fixture.go
Comment (kian99) [reply internal/testutils/jimmtest/fixture.go:35]: Agreed, but that's a refactoring for a separate PR.

**Comment by kian99** [comment]
@alesstimec Addressed your comments, ptal.

File: internal/jimm/jimm.go
Comment (ale8k) [inline internal/jimm/jimm.go:118]: I thought we weren't doing interfaces for now? If we are, can you add one for roles too please.

File: internal/jimm/jimm.go
Comment (ale8k) [inline internal/jimm/jimm.go:160]: Don't start getters with "Get", it is "neither idiomatic nor necessary" according to effective Go:
https://go.dev/doc/effective_go#Getters

File: internal/jimmhttp/rebac_admin/groups_test.go
Comment (ale8k) [inline internal/jimmhttp/rebac_admin/groups_test.go]: Mock is still called GroupService, let's rename it so this doesn't end up really confusing down the line?

**Review by ale8k — CHANGES_REQUESTED** [review]
Some bits and bob needs sorting first, then lgtm.

File: internal/jimm/jimm.go
Comment (ale8k) [reply internal/jimm/jimm.go:118]: Nvm gotten confused

File: internal/jimm/jimm.go
Comment (kian99) [reply internal/jimm/jimm.go:160]: It can't be named GroupManager because there is a field named GroupManager. Once we get rid of the large jimm interface, we can delete these getters.

File: internal/jimmhttp/rebac_admin/groups_test.go
Comment (kian99) [reply internal/jimmhttp/rebac_admin/groups_test.go]: Sure, to `groupMock` or `groupManagerMock`?

Comment (kian99) [reply internal/jimmhttp/rebac_admin/groups_test.go]: I'll go with just GroupManager since it's in the mock package it should be clear.

File: internal/jimm/group/group.go
Comment (alesstimec) [inline internal/jimm/group/group.go]: error message is wrong

File: internal/jimm/group/group.go
Comment (alesstimec) [inline internal/jimm/group/group.go]: error message is wrong

**Review by alesstimec — APPROVED** [review]
LGTM but the error messages in the constructor need changing

File: internal/jimm/group/group.go
Comment (kian99) [reply internal/jimm/group/group.go]: Thanks, fixed.

File: internal/jujuapi/access_control.go
Comment (alesstimec) [inline internal/jujuapi/access_control.go:36]: what if 
```
r.jimm.GetGroupManager()
```
returns nil?

Comment (kian99) [reply internal/jujuapi/access_control.go:36]: I consider that the same as if the `r.jimm.db` is nil or `r.jimm.openFGAClient` is nil. Just let it panic because at that point we messed up. For now, that's more possible to happen because we don't have a jimm constructor but once your PR lands then we have a stronger guarantee this can't be nil.

Comment (alesstimec) [reply internal/jujuapi/access_control.go:36]: ok. agreed. these scenarios should be caught by constructors. same way `r.jimm` could be nil, right...

File: internal/jimm/service_account_test.go
Comment (SimoneDutto) [inline internal/jimm/service_account_test.go:249]: i don't particularly like this, but i guess it's solved by ales's new constructor

File: internal/jimmhttp/rebac_admin/groups.go
Comment (SimoneDutto) [inline internal/jimmhttp/rebac_admin/groups.go:1]: can we change the groupService constructor to accept the `GroupManager` interface?
So we can get rid of the `jujuapi.JIMM` interface.

**Review by SimoneDutto — COMMENTED** [review]
one comment

File: internal/jimm/service_account_test.go
Comment (ale8k) [reply internal/jimm/service_account_test.go:249]: Ya, this be solved

File: internal/testutils/jimmtest/fixture.go
Comment (ale8k) [reply internal/testutils/jimmtest/fixture.go:35]: ALso this was temporary, wasn't supposed to be a catchall function when I did it but we ended up blasting it everywhere aha. We need a better way of doing this really

**Review by ale8k — APPROVED** [review]
lgtm,

File: internal/jimmhttp/rebac_admin/groups.go
Comment (kian99) [reply internal/jimmhttp/rebac_admin/groups.go:1]: We can't yet because the group service also does things like `s.jimm.ListRelationshipTuples` and `s.jimm.AddRelation`. At a later stage we can pass in just the managers that are required.

**Pull request merged**
